go-selinux/selinux_linux.go

```diff
@@ -445,6 +460,186 @@ func ComputeCreateContext(source string, target string, class string) (string, e
+	bitset := &bitset.BitSet{}
+
+	catlist := strings.Split(cats, ",")
+	for _, r := range catlist {
+		ranges := strings.Split(r, ".")
+		if len(ranges) > 1 {
+			catstart, err := strconv.Atoi(string([]rune(ranges[0])[1:]))
+			if err != nil {
+				return nil, err
+			}
+			catend, err := strconv.Atoi(string([]rune(ranges[1])[1:]))
+			if err != nil {
+				return nil, err
+			}
+			for i := catstart; i <= catend; i++ {
+				bitset.Set(uint(i))
+			}
+		} else {
+			cat, err := strconv.Atoi(string([]rune(ranges[0])[1:]))
+			if err != nil {
+				return nil, err
+			}
+			bitset.Set(uint(cat))
+		}
+	}
+
+	return bitset, nil
+}
+
+/*
+rangeStrToMLSRange marshals a string representation of a range.
+*/
+func rangeStrToMLSRange(rangeStr string) (*mlsRange, error) {
+	mlsRange := &mlsRange{}
+	levelSlice := strings.Split(rangeStr, "-")
+
+	switch len(levelSlice) {
+	case 2:
+		high := strings.Split(levelSlice[1], ":")
+		sensHigh, err := strconv.Atoi(string([]rune(high[0])[1:]))
+		if err != nil {
+			return nil, err
+		}
+		mlsRange.high = &level{sens: uint32(sensHigh)}
+		if len(high) > 1 {
+			catsHigh, err := catsToBitset(high[1])
+			if err != nil {
+				return nil, err
+			}
+			mlsRange.high.cats = catsHigh
+		}
+		fallthrough
+	case 1:
+		low := strings.Split(levelSlice[0], ":")
+		sensLow, err := strconv.Atoi(string([]rune(low[0])[1:]))
+		if err != nil {
+			return nil, err
+		}
+		mlsRange.low = &level{sens: uint32(sensLow)}
+		if len(low) > 1 {
+			catsLow, err := catsToBitset(low[1])
+			if err != nil {
+				return nil, err
+			}
+			mlsRange.low.cats = catsLow
+		}
+	}
+
+	if mlsRange.high == nil {
+		mlsRange.high = mlsRange.low
+	}
+
+	return mlsRange, nil
+}
+
+/*
+bitsetToStr takes a category bitset and returns it in the
+canonical selinux syntax
+*/
+func bitsetToStr(c *bitset.BitSet) string {
+	var catBuff bytes.Buffer
+	i, e := c.NextSet(0)
+	len := 0
+	for e {
+		if len == 0 {
+			if catBuff.Len() > 0 {
+				catBuff.WriteString(",")
+			}
+			catBuff.WriteString("c" + strconv.FormatInt(int64(i), 10))
+		}
+
+		next, e := c.NextSet(i + 1)
+		if e {
+			// consecutive cats
+			if next == i+1 {
+				len++
+				i = next
+				continue
+			}
+		}
+		if len == 1 {
+			catBuff.WriteString(",c" + strconv.FormatInt(int64(i), 10))
+		} else if len > 1 {
+			catBuff.WriteString(".c" + strconv.FormatInt(int64(i), 10))
+		}
+		if !e {
+			break
+		}
+		len = 0
+		i = next
+	}
+
+	return catBuff.String()
+}
+
+/*
+String returns an mlsRange as a string.
+*/
+func (m mlsRange) String() string {
+	low := fmt.Sprintf("s%d", m.low.sens)
+	if m.low.cats != nil && m.low.cats.Count() > 0 {
+		low = fmt.Sprintf("%s:%s", low, bitsetToStr(m.low.cats))
+	}
+
+	if reflect.DeepEqual(m.low, m.high) {
+		return low
+	}
+
+	high := fmt.Sprintf("s%d", m.high.sens)
+	if m.high.cats != nil && m.high.cats.Count() > 0 {
+		high = fmt.Sprintf("%s:%s", high, bitsetToStr(m.high.cats))
+	}
+
+	return fmt.Sprintf("%s-%s", low, high)
```

Comment:
nit: this one (and a few `Sprintf`s above can be mere string concatenations, e.g. `low + "-" + high`. Printf is slow in Go.